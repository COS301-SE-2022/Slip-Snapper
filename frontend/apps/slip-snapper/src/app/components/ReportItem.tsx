import {
  IonButton,
  IonCard,
  IonCardHeader,
  IonCardSubtitle,
  IonCardTitle,
  IonCol,
  IonItem,
} from '@ionic/react';
import { isPlatform } from '@ionic/core';
import { Filesystem, Directory, Encoding } from '@capacitor/filesystem';
import { FileOpener } from '@ionic-native/file-opener';
import { getUserReport } from '../../api/apiCall';
import '../theme/reportItem.css';

type Props = {
  reportData: string[];
};
function ReportItem({ reportData }: Props) {
  return (
    <IonCol className="recent-item-col">
      <IonCard color="primary">
        <IonCardHeader>
          <IonCardSubtitle>Report {reportData[0]} :</IonCardSubtitle>
          <IonCardTitle>{reportData[1]}</IonCardTitle>
        </IonCardHeader>
        <IonItem color="tertiary">
          <IonButton
            onClick={() => view(reportData[1])}
            fill="solid"
            slot="end"
            color="secondary"
          >
            View
          </IonButton>
        </IonItem>
      </IonCard>
    </IonCol>
  );
}
function view(data: any) {
  let user = JSON.parse(localStorage.getItem('user')!);
  if (user == null) {
    user = { username: 'demoUser' };
  }
  getUserReport(user.username, data).then((apiResponse) => {
    if (apiResponse.data.report.data !== undefined) {
      const arr = new Uint8Array(apiResponse.data.report.data);
      const blob = new Blob([arr], { type: 'application/pdf' });
      const docUrl = URL.createObjectURL(blob);

      if (!isPlatform('android') && !isPlatform('ios')) {
        window.open(docUrl);
      } else {
        //view for mobile, might need name
        const reader = new FileReader();

        reader.addEventListener(
          'load',
          () => {
            if (reader.result) {
              const result = reader.result as string;
              const pdfData = result.split(',')[1];
              downloadPDF(pdfData);
            }
          },
          false
        );

        reader.readAsDataURL(blob);
      }
    }
  });
}

function downloadPDF(pdfBase64: string) {
  try {
    Filesystem.writeFile({
      path: 'report.pdf',
      data: pdfBase64,
      directory: Directory.External,
    }).then((writeFileResult) => {
      Filesystem.getUri({
        directory: Directory.External,
        path: 'report.pdf',
      }).then(
        (getUriResult) => {
          const path = getUriResult.uri;
          FileOpener.open(path, 'application/pdf').then(() =>
            console.log('File is opened')
          );
        },
        (error) => {
          console.log(error);
        }
      );
    });
  } catch (error) {
    console.error('Unable to write file', error);
  }
}
export default ReportItem;
