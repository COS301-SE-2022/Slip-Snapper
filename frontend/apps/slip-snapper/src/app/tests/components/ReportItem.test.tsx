import renderer from 'react-test-renderer';
import ReportItem from '../../components/ReportItem';

jest.mock('@capacitor/filesystem');
jest.mock('../../api/apiCall');
jest.mock('@ionic-native/file-opener/index', () => jest.fn());

const renderTree = (tree: JSX.Element) => renderer.create(tree);
describe('<ReportItem>', () => {
  it('should render component with props', () => {
    expect(renderTree(<ReportItem  
      reportData={["18", '19-9-2022_Weekly_18.pdf', '19/9/2022 Weekly #18.pdf']} 
    />).toJSON()).toMatchSnapshot();
  });
});