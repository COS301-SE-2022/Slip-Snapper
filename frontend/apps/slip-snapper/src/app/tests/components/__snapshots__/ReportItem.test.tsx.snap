// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportItem> should render component with props 1`] = `
<ion-col>
  <ion-card
    color="primary"
    onClick={[Function]}
  >
    <ion-card-header>
      <ion-card-title>
        Report 
        18
        :
      </ion-card-title>
      <ion-text>
        19/9/2022 Weekly #18.pdf
      </ion-text>
    </ion-card-header>
    <ion-item
      color="tertiary"
      onClick={[Function]}
    >
      <ion-button
        color="secondary"
        fill="solid"
        onClick={[Function]}
        slot="end"
      >
        View
      </ion-button>
    </ion-item>
  </ion-card>
</ion-col>
`;
